File: src/main/java/com/tibiawiki/process/RetrieveCharms.java
package com.tibiawiki.process;

import com.tibiawiki.domain.enums.InfoboxTemplate;
import com.tibiawiki.domain.factories.ArticleFactory;
import com.tibiawiki.domain.factories.JsonFactory;
import com.tibiawiki.domain.repositories.ArticleRepository;
import org.json.JSONObject;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@Component
public class RetrieveCharms extends RetrieveAny {

    @Autowired
    public RetrieveCharms(ArticleRepository articleRepository, ArticleFactory articleFactory, JsonFactory jsonFactory) {
        super(articleRepository, articleFactory, jsonFactory);
    }

    public List<String> getCharmsList() {
        final List<String> charmsCategory = articleRepository.getPageNamesFromCategory(InfoboxTemplate.CHARM.getCategoryName());
        final List<String> listsCategory = articleRepository.getPageNamesFromCategory(CATEGORY_LISTS);

        return charmsCategory.stream()
                .filter(page -> !listsCategory.contains(page))
                .collect(Collectors.toList());
    }

    public Stream<JSONObject> getCharmsJSON() {
        return getArticlesFromInfoboxTemplateAsJSON(getCharmsList());
    }

    public Optional<JSONObject> getCharmJSON(String pageName) {
        return super.getArticleAsJSON(pageName);
    }
}

File: src/main/kotlin/com/tibiawiki/process/RetrieveCharms.kt
package com.tibiawiki.process

import com.tibiawiki.domain.WikiJson
import com.tibiawiki.domain.enums.InfoboxTemplate
import com.tibiawiki.domain.factories.ArticleFactory
import com.tibiawiki.domain.factories.JsonFactory
import com.tibiawiki.domain.repositories.ArticleRepository
import org.springframework.stereotype.Component
import java.util.Optional
import java.util.stream.Stream

@Component
class RetrieveCharms(
    articleRepository: ArticleRepository,
    articleFactory: ArticleFactory,
    jsonFactory: JsonFactory
) : RetrieveAny(articleRepository, articleFactory, jsonFactory) {

    val charmsList: List<String>
        get() {
            val category = articleRepository.getPageNamesFromCategory(InfoboxTemplate.CHARM.categoryName)
            val listsCategory = articleRepository.getPageNamesFromCategory(CATEGORY_LISTS)
            return category.filter { page -> page !in listsCategory }
        }

    val charmsJSON: Stream<WikiJson>
        get() = getArticlesFromInfoboxTemplateAsJSON(charmsList)

    fun getCharmJSON(pageName: String): Optional<WikiJson> {
        return getArticleAsJSON(pageName)
    }
}
